shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/impl/MySQLDDLVisitor.java

```diff
@@ -461,92 +483,256 @@ public ASTNode visitKeyParts_(final KeyParts_Context ctx) {
+        CollectionValue<ValidStatementSegment> result = new CollectionValue<>();
+        ValidStatementSegment validStatement = new ValidStatementSegment(ctx.start.getStartIndex(), ctx.stop.getStopIndex());
+        if (null != ctx.createTable()) {
+            validStatement.setCreateTable((MySQLCreateTableStatement) visit(ctx.createTable()));
+            result.getValue().add(validStatement);
+        } else if (null != ctx.alterTable()) {
+            validStatement.setAlterTable((MySQLAlterTableStatement) visit(ctx.alterTable()));
+            result.getValue().add(validStatement);
+        } else if (null != ctx.dropTable()) {
+            validStatement.setDropTable((MySQLDropTableStatement) visit(ctx.dropTable()));
+            result.getValue().add(validStatement);
+        } else if (null != ctx.truncateTable()) {
+            validStatement.setTruncate((MySQLTruncateStatement) visit(ctx.truncateTable()));
+            result.getValue().add(validStatement);
+        }
+        if (null != ctx.validDMLStatement()) {
+            result.combine((CollectionValue<ValidStatementSegment>) visit(ctx.validDMLStatement()));
+        }
+        if (null != ctx.beginStatement()) {
+            result.combine((CollectionValue<ValidStatementSegment>) visit(ctx.beginStatement()));
+        }
+        if (null != ctx.flowControlStatement()) {
+            result.combine((CollectionValue<ValidStatementSegment>) visit(ctx.flowControlStatement()));
+        }
+        return result;
+    }
+    
+    @Override
+    public ASTNode visitValidDMLStatement(final ValidDMLStatementContext ctx) {
+        CollectionValue<ValidStatementSegment> result = new CollectionValue<>();
+        ValidStatementSegment validStatement = new ValidStatementSegment(ctx.start.getStartIndex(), ctx.stop.getStopIndex());
+        if (null != ctx.insert()) {
+            validStatement.setInsert((MySQLInsertStatement) createParseTreeVisitor(ctx.insert().getClass()).visit(ctx.insert()));
+        } else if (null != ctx.replace()) {
+            validStatement.setReplace((MySQLInsertStatement) createParseTreeVisitor(ctx.replace().getClass()).visit(ctx.replace()));
+        } else if (null != ctx.update()) {
+            validStatement.setUpdate((MySQLUpdateStatement) createParseTreeVisitor(ctx.update().getClass()).visit(ctx.update()));
+        } else if (null != ctx.delete()) {
+            validStatement.setDelete((MySQLDeleteStatement) createParseTreeVisitor(ctx.delete().getClass()).visit(ctx.delete()));
+        } else if (null != ctx.select()) {
+            validStatement.setSelect((MySQLSelectStatement) createParseTreeVisitor(ctx.select().getClass()).visit(ctx.select()));
+        }
+        result.getValue().add(validStatement);
+        return result;
+    }
+    
+    @SuppressWarnings("rawtypes")
+    private ParseTreeVisitor createParseTreeVisitor(final Class<? extends ParseTree> parseTreeClass) {
```

Comment:
> if `MySQLDMLVisitor` is necessary, do you think moving the content of `MySQLDMLVisitor` to `MySQLVisitor` will be better than `createParseTreeVisitor`?

@tristaZero The stored routine supports a variety of DML and DAL statements, and migrating the implementation of these statements to the `MySQLVisitor` would cause the `MySQLVisitor` to become bloated. 
In addition, the `MySQLVisitor` seems to be more suited to processing some generic syntax fragments than the entire statement.

Reply:
Yep, I buy that  `MySQLVisitor` is supposed to contain the generic functions and items. But if `MySQLDDLVisitor` needs to use these functions of `MySQLDMLVisitor`,  the relevant ones will become shared one for `DMLVisitor` and `DDLVisitor`, and is possible to move them to `MySQLVisitor`, though which will create a bloated father class.

What's more, as you know, the `currentParameterIndex` makes `MySQLVisitor` not as a simple utility. If you create a new instance of `DMLVisitor` in `DDLVisitor`, I guess there will be an incorrect parameterIndex.

Honestly speaking, I have no better method to handle this issue. :-(
Actually, for some of the other SQL parser engines like `druid` or `hive`, they put all the visitors and g4 files together, as they do not have a complete and an awful lot of SQL definitions. 😅

Reply:
@tristaZero Thank you very much for your advice, the setting of parameter `currentParameterIndex` is really a problem, I will optimize it by migrating the common function from the MySQLDMLVisitor to the MySQLVisitor. 😀